src/transformStateWithClones.js

```
@@ -7,7 +7,26 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  const states = [];
+  let preState = { ...state };
+
+  for (const action of actions) {
+    let newState = { ...preState };
+
+    if (action.type === 'addProperties') {
+      newState = { ...preState, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      for (const key of action.keysToRemove) {
+        delete newState[key];
+      }
```

This does not validate `keysToRemove` for the `removeProperties` action. The task description requires: "`removeProperties` - remove all keys given in the `keysToRemove` array from the `state`. (ignore not existing)" The loop `for (const key of action.keysToRemove)` (line 19) assumes `keysToRemove` is present and iterable; if it's undefined this will throw. Check `Array.isArray(action.keysToRemove)` (or otherwise verify iterability) before looping and ensure non-existing keys are silently ignored as required .